# Aede local type-safety patch: unsigned 32-bit bitstream word assembly.
# BSD copyright/licence retained; no decoder algorithm or PCM behavior change.
--- a/libmpcdec/mpc_bits_reader.h
+++ b/libmpcdec/mpc_bits_reader.h
@@ -53,12 +53,12 @@
 
 	ret = (r->buff[0] | (r->buff[-1] << 8)) >> r->count;
 	if (nb_bits > (16 - r->count)) {
-		ret |= (mpc_uint32_t)((r->buff[-2] << 16) | (r->buff[-3] << 24)) >> r->count;
+		ret |= (mpc_uint32_t)(((mpc_uint32_t)r->buff[-2] << 16) | ((mpc_uint32_t)r->buff[-3] << 24)) >> r->count;
 		if (nb_bits > 24 && r->count != 0)
-			ret |= r->buff[-4] << (32 - r->count);
+			ret |= (mpc_uint32_t)r->buff[-4] << (32 - r->count);
 	}
 
-	return ret & ((1 << nb_bits) - 1);
+	return ret & (((mpc_uint32_t)1 << nb_bits) - 1);
 }
 
 // basic huffman decoding routine
